python/geom.py

```diff
@@ -237,31 +240,37 @@ def transform(self, m):
             s.transform(m)
 
     def epsilon(self,freq):
-
+        return self._get_epsmu(self.epsilon_diag, self.epsilon_offdiag, self.E_susceptibilities, self.D_conductivity_diag, self.D_conductivity_offdiag, freq)
+    
+    def mu(self,freq):
+        return self._get_epsmu(self.mu_diag, self.mu_offdiag, self.H_susceptibilities, self.B_conductivity_diag, self.B_conductivity_offdiag, freq)
+    
+    def _get_epsmu(self, diag, offdiag, susceptibilities, conductivity_diag, conductivity_offdiag, freq):
         # Clean the input
-        if not isinstance(freq, list):
-            raise ValueError('Frequencies must be supplied as a list')
+        if type(freq) is not np.ndarray:
+            freqs = np.array(freq)
+        else:
+            freqs = np.squeeze(freq)
+
+        numFreqs = freqs.size
 
         # Initialize with instantaneous dielectric tensor
-        eps = [Matrix(mp.Vector3(self.epsilon_diag.x, self.epsilon_offdiag.x, self.epsilon_offdiag.y),
-                     mp.Vector3(self.epsilon_offdiag.x, self.epsilon_diag.y, self.epsilon_offdiag.z),
-                     mp.Vector3(self.epsilon_offdiag.y, self.epsilon_offdiag.z, self.epsilon_diag.z)) for i in range(len(freq))]
-
-        # Iterate through frequencies
-        for i_freq in range(len(freq)):
-            # Iterate through susceptibilities
-            for i_sus in range(len(self.E_susceptibilities)):
-                eps[i_freq] = eps[i_freq] + self.E_susceptibilities[i_sus].eval_susceptibility(freq[i_freq])
-            # Account for conductivity term
-            D_conductivity = Matrix(mp.Vector3(self.D_conductivity_diag.x, self.D_conductivity_offdiag.x, self.D_conductivity_offdiag.y),
-                       mp.Vector3(self.D_conductivity_offdiag.x, self.D_conductivity_diag.y, self.D_conductivity_offdiag.z),
-                       mp.Vector3(self.D_conductivity_offdiag.y, self.D_conductivity_offdiag.z, self.D_conductivity_diag.z))
-            ones_matrix = Matrix(mp.Vector3(1, 1, 1),
-                       mp.Vector3(1, 1, 1),
-                       mp.Vector3(1, 1, 1))
-            eps[i_freq] = (ones_matrix + 1j/freq[i_freq] * D_conductivity) * eps[i_freq]
+        epsmu = np.tile(np.array(Matrix(mp.Vector3(diag.x, offdiag.x, offdiag.y),
+                     mp.Vector3(offdiag.x, diag.y, offdiag.z),
+                     mp.Vector3(offdiag.y, offdiag.z, diag.z))),(numFreqs,1,1))
```

Comment:
Shouldn't this be a `(numFreqs,3,3)` array of 3x3 matrices, rather than an array of `Matrix`?   (Probably we need t a utility function to make a 3x3 hermitian numpy matrix from `diag` and `offdiag`.)

Also, I think you can just make it of size `(1,3,3)` and not use `np.tile` here, since when you multiply it by a `(numfreqs, 3,3)` array the numpy broadcasting will implicitly do the tiling for you.

(You should hardly ever need to call `np.tile` because of broadcasting, in fact.)

Reply:
Actually, the `Matrix` class has an `__array__` overload function that automatically converts it into a numpy array when wrapped by the `np.array()` function. I just use the `Matrix` class for readability.

I agree, however, that we could include a utility function for these types of Matrix initializations. 

I'll also leverage the broadcasting better. I'll have to add an additional dimension to all of the `Matrix` objects to convert it from a 3x3 `numpy` array to a 1x3x3. You're right, there's no reason to use `tile` if I set up all the arrays right.